### PR TITLE
GRAPHICS: Shader material custom blending

### PR DESCRIPTION
**Pull request by mirv-sillyfish**
Explicitly specify blending equations and functions if non-default
blending is used.

### CODE REVIEW

File: src/graphics/shader/shadermaterial.h [hunk 1]
```diff
@@ -38,7 +38,7 @@ class ShaderMaterial {
 		MATERIAL_TRANSPARENT   = 0x0002,  ///< Material definitely has transparency.
 		MATERIAL_TRANSPARENT_B = 0x0004,  ///< Material definitely has transparency.
 		MATERIAL_DECAL         = 0x0008,  ///< Material is a decal.
-		MATERIAL_SPECIAL_BLEND = 0x0010   ///< Different to default blending. Maybe this should be extended one day.
+		MATERIAL_CUSTOM_BLEND  = 0x0010,  ///< Different blending is being used, see _blendMode.
```
Comment (mirv-sillyfish) [inline src/graphics/shader/shadermaterial.h]: Argh, note to self: bad comment on variable name. Will clean up after after comments are added.

File: src/graphics/aurora/guiquad.cpp [hunk 2]
```diff
@@ -297,6 +310,74 @@ void GUIQuad::render(RenderPass pass) {
 	glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
 }
 
+void GUIQuad::renderImmediate(const glm::mat4 &parentTransform)
+{
```
Comment (DrMcCoy) [inline src/graphics/aurora/guiquad.cpp]: That brace needs to go onto the previous line

File: src/graphics/aurora/guiquad.cpp [hunk 3]
```diff
@@ -297,6 +310,74 @@ void GUIQuad::render(RenderPass pass) {
...
+	if (_xor) {
+		glDisable(GL_COLOR_LOGIC_OP);
+	}
+
+	if (_scissor)
+		glDisable(GL_SCISSOR_TEST);
+}
+
+void GUIQuad::buildMaterial()
+{
```
Comment (DrMcCoy) [inline src/graphics/aurora/guiquad.cpp]: Here too

**Comment by DrMcCoy** [comment]
I get a sanitizer error in KotOR (without the new render path):

```
=================================================================
==20145==ERROR: AddressSanitizer: heap-use-after-free on address 0x60b00009c258 at pc 0x56236ee25307 bp 0x7f78555bfc00 sp 0x7f78555bfbf0
READ of size 4 at 0x60b00009c258 thread T8 (GameInstance)
    #0 0x56236ee25306 in Graphics::Shader::ShaderMaterial::useDecrement() /home/drmccoy/projects/xoreos/xoreos/src/graphics/shader/shadermaterial.cpp:412
    #1 0x56236ee2b094 in Graphics::Shader::ShaderRenderable::setMaterial(Graphics::Shader::ShaderMaterial*, bool) /home/drmccoy/projects/xoreos/xoreos/src/graphics/shader/shaderrenderable.cpp:117
    #2 0x56236ee2b2ad in Graphics::Shader::ShaderRenderable::operator=(Graphics::Shader::ShaderRenderable const&) /home/drmccoy/projects/xoreos/xoreos/src/graphics/shader/shaderrenderable.cpp:67
    #3 0x56236e95eb2f in Graphics::Aurora::GUIQuad::buildMaterial() /home/drmccoy/projects/xoreos/xoreos/src/graphics/aurora/guiquad.cpp:368
    #4 0x56236e96355b in Graphics::Aurora::GUIQuad::setTexture(Common::UString const&) /home/drmccoy/projects/xoreos/xoreos/src/graphics/aurora/guiquad.cpp:180
    #5 0x56236e32f88c in Engines::KotORJadeWidget::setFill(Common::UString const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/kotorjadegui/kotorjadewidget.cpp:198
    #6 0x56236dc1ced9 in Engines::KotOR::LoadScreen::LoadScreen(Common::UString const&, Engines::Console*) /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/loadscreen/loadscreen.cpp:44
    #7 0x56236db2f4fe in Engines::KotOR::MainMenu::createClassSelection() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/main/main.cpp:97
    #8 0x56236db35042 in Engines::KotOR::MainMenu::callbackActive(Engines::Widget&) /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/main/main.cpp:190
    #9 0x56236e304fd9 in Engines::GUI::checkWidgetActive(Engines::Widget*) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:492
    #10 0x56236e30d30e in Engines::GUI::mouseUp(SDL_Event const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:539
    #11 0x56236e30f624 in Engines::GUI::mouseUp(SDL_Event const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:526
    #12 0x56236e30f624 in Engines::GUI::processEventQueue() /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:204
    #13 0x56236e315f7c in Engines::GUI::run(unsigned int) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:132
    #14 0x56236d9fce69 in Engines::KotOR::Game::mainMenu() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/game.cpp:150
    #15 0x56236d9fdaff in Engines::KotOR::Game::run() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/game.cpp:75
    #16 0x56236d9dd290 in Engines::KotOR::KotOREngine::run() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/kotor.cpp:128
    #17 0x56236e2758c6 in Engines::GameInstanceEngine::run() /home/drmccoy/projects/xoreos/xoreos/src/engines/enginemanager.cpp:225
    #18 0x56236e277760 in Engines::EngineManager::run(Engines::GameInstance&) const /home/drmccoy/projects/xoreos/xoreos/src/engines/enginemanager.cpp:252
    #19 0x56236e27912b in Engines::GameThread::threadMethod() /home/drmccoy/projects/xoreos/xoreos/src/engines/gamethread.cpp:87
    #20 0x56236f3d8233 in Common::Thread::threadHelper(void*) /home/drmccoy/projects/xoreos/xoreos/src/common/thread.cpp:113
    #21 0x7f787702f18b in SDL_RunThread /var/tmp/portage/media-libs/libsdl2-2.0.9/work/SDL2-2.0.9/src/thread/SDL_thread.c:283
    #22 0x7f7877096478 in RunThread /var/tmp/portage/media-libs/libsdl2-2.0.9/work/SDL2-2.0.9/src/thread/pthread/SDL_systhread.c:78
    #23 0x7f7875f3a9a9 in start_thread /var/tmp/portage/sys-libs/glibc-2.27-r6/work/glibc-2.27/nptl/pthread_create.c:463
    #24 0x7f7871bbf56e in clone (/lib64/libc.so.6+0x10456e)

0x60b00009c258 is located 104 bytes inside of 112-byte region [0x60b00009c1f0,0x60b00009c260)
freed by thread T8 (GameInstance) here:
    #0 0x7f7876243738 in operator delete(void*, unsigned long) /var/tmp/portage/sys-devel/gcc-8.2.0-r4/work/gcc-8.2.0/libsanitizer/asan/asan_new_delete.cc:151
    #1 0x56236e95e465 in Graphics::Aurora::GUIQuad::buildMaterial() /home/drmccoy/projects/xoreos/xoreos/src/graphics/aurora/guiquad.cpp:349
    #2 0x56236e96355b in Graphics::Aurora::GUIQuad::setTexture(Common::UString const&) /home/drmccoy/projects/xoreos/xoreos/src/graphics/aurora/guiquad.cpp:180
    #3 0x56236e32f88c in Engines::KotORJadeWidget::setFill(Common::UString const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/kotorjadegui/kotorjadewidget.cpp:198
    #4 0x56236dc1ced9 in Engines::KotOR::LoadScreen::LoadScreen(Common::UString const&, Engines::Console*) /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/loadscreen/loadscreen.cpp:44
    #5 0x56236db2f4fe in Engines::KotOR::MainMenu::createClassSelection() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/main/main.cpp:97
    #6 0x56236db35042 in Engines::KotOR::MainMenu::callbackActive(Engines::Widget&) /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/main/main.cpp:190
    #7 0x56236e304fd9 in Engines::GUI::checkWidgetActive(Engines::Widget*) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:492
    #8 0x56236e30d30e in Engines::GUI::mouseUp(SDL_Event const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:539
    #9 0x56236e30f624 in Engines::GUI::mouseUp(SDL_Event const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:526
    #10 0x56236e30f624 in Engines::GUI::processEventQueue() /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:204
    #11 0x56236e315f7c in Engines::GUI::run(unsigned int) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:132
    #12 0x56236d9fce69 in Engines::KotOR::Game::mainMenu() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/game.cpp:150
    #13 0x56236d9fdaff in Engines::KotOR::Game::run() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/game.cpp:75
    #14 0x56236d9dd290 in Engines::KotOR::KotOREngine::run() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/kotor.cpp:128
    #15 0x56236e2758c6 in Engines::GameInstanceEngine::run() /home/drmccoy/projects/xoreos/xoreos/src/engines/enginemanager.cpp:225
    #16 0x56236e277760 in Engines::EngineManager::run(Engines::GameInstance&) const /home/drmccoy/projects/xoreos/xoreos/src/engines/enginemanager.cpp:252
    #17 0x56236e27912b in Engines::GameThread::threadMethod() /home/drmccoy/projects/xoreos/xoreos/src/engines/gamethread.cpp:87
    #18 0x56236f3d8233 in Common::Thread::threadHelper(void*) /home/drmccoy/projects/xoreos/xoreos/src/common/thread.cpp:113
    #19 0x7f787702f18b in SDL_RunThread /var/tmp/portage/media-libs/libsdl2-2.0.9/work/SDL2-2.0.9/src/thread/SDL_thread.c:283

previously allocated by thread T8 (GameInstance) here:
    #0 0x7f78762421c0 in operator new(unsigned long) /var/tmp/portage/sys-devel/gcc-8.2.0-r4/work/gcc-8.2.0/libsanitizer/asan/asan_new_delete.cc:90
    #1 0x56236e95edb6 in Graphics::Aurora::GUIQuad::buildMaterial() /home/drmccoy/projects/xoreos/xoreos/src/graphics/aurora/guiquad.cpp:361
    #2 0x56236e960a79 in Graphics::Aurora::GUIQuad::GUIQuad(Common::UString const&, float, float, float, float, float, float, float, float) /home/drmccoy/projects/xoreos/xoreos/src/graphics/aurora/guiquad.cpp:69
    #3 0x56236e96f704 in Graphics::Aurora::HighlightableGUIQuad::HighlightableGUIQuad(Common::UString const&, float, float, float, float, float, float, float, float) /home/drmccoy/projects/xoreos/xoreos/src/graphics/aurora/highlightableguiquad.cpp:30
    #4 0x56236e340acd in Engines::KotORJadeWidget::load(Aurora::GFF3Struct const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/kotorjadegui/kotorjadewidget.cpp:123
    #5 0x56236d9ad586 in Engines::KotOR::GUI::createWidget(Engines::KotOR::GUI::WidgetContext&) /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/gui.cpp:181
    #6 0x56236d9ae822 in Engines::KotOR::GUI::loadWidget(Aurora::GFF3Struct const&, Engines::Widget*) /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/gui.cpp:130
    #7 0x56236d9b0cb0 in Engines::KotOR::GUI::load(Common::UString const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/gui.cpp:118
    #8 0x56236dc1cc6e in Engines::KotOR::LoadScreen::LoadScreen(Common::UString const&, Engines::Console*) /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/loadscreen/loadscreen.cpp:42
    #9 0x56236db2f4fe in Engines::KotOR::MainMenu::createClassSelection() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/main/main.cpp:97
    #10 0x56236db35042 in Engines::KotOR::MainMenu::callbackActive(Engines::Widget&) /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/gui/main/main.cpp:190
    #11 0x56236e304fd9 in Engines::GUI::checkWidgetActive(Engines::Widget*) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:492
    #12 0x56236e30d30e in Engines::GUI::mouseUp(SDL_Event const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:539
    #13 0x56236e30f624 in Engines::GUI::mouseUp(SDL_Event const&) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:526
    #14 0x56236e30f624 in Engines::GUI::processEventQueue() /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:204
    #15 0x56236e315f7c in Engines::GUI::run(unsigned int) /home/drmccoy/projects/xoreos/xoreos/src/engines/aurora/gui.cpp:132
    #16 0x56236d9fce69 in Engines::KotOR::Game::mainMenu() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/game.cpp:150
    #17 0x56236d9fdaff in Engines::KotOR::Game::run() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/game.cpp:75
    #18 0x56236d9dd290 in Engines::KotOR::KotOREngine::run() /home/drmccoy/projects/xoreos/xoreos/src/engines/kotor/kotor.cpp:128
    #19 0x56236e2758c6 in Engines::GameInstanceEngine::run() /home/drmccoy/projects/xoreos/xoreos/src/engines/enginemanager.cpp:225
    #20 0x56236e277760 in Engines::EngineManager::run(Engines::GameInstance&) const /home/drmccoy/projects/xoreos/xoreos/src/engines/enginemanager.cpp:252
    #21 0x56236e27912b in Engines::GameThread::threadMethod() /home/drmccoy/projects/xoreos/xoreos/src/engines/gamethread.cpp:87
    #22 0x56236f3d8233 in Common::Thread::threadHelper(void*) /home/drmccoy/projects/xoreos/xoreos/src/common/thread.cpp:113
    #23 0x7f787702f18b in SDL_RunThread /var/tmp/portage/media-libs/libsdl2-2.0.9/work/SDL2-2.0.9/src/thread/SDL_thread.c:283

Thread T8 (GameInstance) created by T0 here:
    #0 0x7f787619db73 in __interceptor_pthread_create /var/tmp/portage/sys-devel/gcc-8.2.0-r4/work/gcc-8.2.0/libsanitizer/asan/asan_interceptors.cc:202
    #1 0x7f7877096546 in SDL_SYS_CreateThread /var/tmp/portage/media-libs/libsdl2-2.0.9/work/SDL2-2.0.9/src/thread/pthread/SDL_systhread.c:119

SUMMARY: AddressSanitizer: heap-use-after-free /home/drmccoy/projects/xoreos/xoreos/src/graphics/shader/shadermaterial.cpp:412 in Graphics::Shader::ShaderMaterial::useDecrement()
Shadow bytes around the buggy address:
  0x0c168000b7f0: 00 00 00 fa fa fa fa fa fa fa fa fa 00 00 00 00
  0x0c168000b800: 00 00 00 00 00 00 00 00 00 00 fa fa fa fa fa fa
  0x0c168000b810: fa fa 00 00 00 00 00 00 00 00 00 00 00 00 00 fa
  0x0c168000b820: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c168000b830: 00 00 00 00 00 00 fa fa fa fa fa fa fa fa fd fd
=>0x0c168000b840: fd fd fd fd fd fd fd fd fd fd fd[fd]fa fa fa fa
  0x0c168000b850: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c168000b860: 00 00 fa fa fa fa fa fa fa fa 00 00 00 00 00 00
  0x0c168000b870: 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa fa
  0x0c168000b880: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fa fa
  0x0c168000b890: fa fa fa fa fa fa 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==20145==ABORTING
```

**Comment by DrMcCoy** [comment]
Merged as 4e0a98ebcce5...baf0dcf11cf6, thanks! :)

**Pull request closed without merging**